pkg/controller/sidecarterminator/sidecar_terminator_controller.go

```diff
@@ -152,23 +162,137 @@ func (r *ReconcileSidecarTerminator) doReconcile(pod *corev1.Pod) (reconcile.Res
 		return reconcile.Result{}, err
 	}
 
+	if err := r.terminateJobPod(pod, sidecarNeedToSyncStatus); err != nil {
+		return reconcile.Result{}, err
+	}
+
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileSidecarTerminator) groupSidecars(pod *corev1.Pod) (sets.String, sets.String, error) {
+// terminateJobPod terminate the job pod and skip the state of the sidecar containers
+// This method should only be called after the executeKillContainerAction is called
+func (r *ReconcileSidecarTerminator) terminateJobPod(pod *corev1.Pod, sidecars sets.String) error {
+	if sidecars.Len() == 0 {
+		return nil
+	}
+	// skip sync status of sidecar container if job has completed.
+	// the real status that reported by kubelet will be store into the state of sidecar container.
+	// the pod is repeatedly processed by job controller until the job reaches completed phase. because kubelet and the logic bellow will report different status of the sidecar container.
+	if deduceWhetherTheJobIsCompletedFromThePod(pod) {
+		klog.V(3).Infof("SidecarTerminator -- we can deduce whether the job is completed from the main container status of the pod(%v/%v) and pod phase,no need to process", pod.Namespace, pod.Name)
+		return nil
+	}
+
+	var changed bool
+	newSidecarStatus := make(map[string]corev1.ContainerStatus)
+	for i := range pod.Spec.Containers {
+		status := &pod.Status.ContainerStatuses[i]
+		if !sidecars.Has(status.Name) {
+			continue
+		}
+
+		changed = true
+
+		// The kubelet may update the state of the containers and the phase of the pod before the sidecar terminator controller
+		if status.State.Terminated != nil && status.State.Terminated.ExitCode != int32(0) {
+			klog.V(3).Infof("SidecarTerminator -- ignore the non-zero exit code of the sidecar container %s/%s, the pod phase will be updated", pod.Name, status.Name)
+			newSidecarStatus[status.Name] = *status
+		} else if status.State.Terminated == nil && status.State.Running != nil {
+			klog.V(3).Infof("SidecarTerminator -- sync the status of the sidecar container %s/%s and update the pod phase, all of the main containers are completed", pod.Name, status.Name)
```

Comment:
before adding a Terminated  state, i think sidecar terminator should wait for the sidecar container to exit with non zero code

Reply:
In the scenario where kruise-daemon is down, this means that we cannot complete the job until  kruise-daemon is successfully started.

Reply:
maybe we should not care the status of the sidecar containers.  And update the pod phase if no waiting sidecar container.

Reply:
I think you are correct , it will avoid kubelet and sidecar terminator from updating container status unnecessarily.